lib/model/query/entities.js

```diff
@@ -90,38 +90,29 @@ createNew.audit.withResult = true;
 // it could be used in places of createNew() but createNew uses a single query so it may be faster
 // in single entity situations (eg. processing submissions to make entities)
 // Note: if the entity schema changes, createMany and createNew would both need to change.
-const createMany = (dataset, entities, sourceId, userAgentIn) => async ({ all, context }) => {
+const createMany = (dataset, rawEntities, sourceId, userAgentIn) => async ({ all, context }) => {
   const creatorId = context.auth.actor.map((actor) => actor.id).orNull();
   const userAgent = blankStringToNull(userAgentIn);
 
-  const entityInsert = sql.join(entities.map(e => sql`(${sql.join([dataset.id, e.uuid, creatorId, sql`clock_timestamp()`], sql`,`)} )`), sql`,`);
-  const newEntities = await all(sql`
-  INSERT INTO entities ("datasetId", "uuid", "creatorId", "createdAt")
-  VALUES ${entityInsert}
-  RETURNING id`);
-
-  const defInsert = sql.join(entities.map((e, i) => sql`(${sql.join([
-    newEntities[i].id,
-    e.def.label,
-    JSON.stringify(e.def.data),
-    JSON.stringify(e.def.dataReceived),
-    sourceId,
+  // Augment parsed entity data with dataset and creator IDs
+  const entitiesForInsert = rawEntities.map(e => new Entity({ datasetId: dataset.id, creatorId, ...e }));
+
+  const entities = await all(insertMany(entitiesForInsert));
+
+  // Augment defs with IDs of freshly inserted entities and
+  // other default values
+  const defsForInsert = rawEntities.map((e, i) => new Entity.Def({
+    entityId: entities[i].id,
     creatorId,
+    root: true,
+    current: true,
+    sourceId,
+    version: 1,
     userAgent,
-    'true',
-    'true',
-    sql`clock_timestamp()`,
-    '1'
-  ], sql`,`)} )`), sql`,`);
-
-  const defs = await all(sql`
-  INSERT INTO entity_defs ("entityId", "label", "data", "dataReceived",
-    "sourceId", "creatorId", "userAgent", "root", "current", "createdAt", "version")
-  VALUES ${defInsert}
-  RETURNING *
-  `);
-
-  return defs;
+    ...e.def,
+    data: { ...e.def.data }, // was [Object: null prototype]
```

Comment:
Could you say more about this line? Were you encountering an error involving `data`? I think it's a good thing that we create `e.def.data` with a `null` prototype. If we could avoid copying the data, that seems ideal.

Reply:
I was getting this error:
```
'InvalidInputError: Invalid unnest tuple member type. Must be a primitive value expression.',
'at Object.createUnnestSqlFragment (/Users/ktuite/Desktop/code/odk/central-backend/node_modules/slonik/dist/src/sqlFragmentFactories/createUnnestSqlFragment.js:35:23)',
...
```

Reply:
Hmm it sort of sounds like `JSON.stringify()` isn't getting called. 🤔 I'm guessing that this function of ours is at fault:

https://github.com/getodk/central-backend/blob/7839e07b88facbbf10336cfdd38ce3d9242405f7/lib/util/db.js#L135-L143

If that's the cause, I think we could call `JSON.stringify()` there if `typeof v === 'object' && Object.getPrototypeOf(v) == null`. I'm not sure how best to modify that ternary, but I'm thinking of something like:

```js
const _assign = (obj) => (k) => {
  if (k === 'createdAt') return sql`clock_timestamp()`;
  const v = obj[k];
  if (v === null || v === undefined) return null;
  if (typeof v === 'object') {
    if (v instanceof Date) return v.toISOString();
    if (v.constructor === Object || Object.getPrototypeOf(v) == null) return JSON.stringify(v);
  }
  return v;
};
```